src/nvim/msgpack_rpc/channel.c

```diff
@@ -521,8 +521,19 @@ void rpc_close(Channel *channel)
   }
 }
 
+static void exit_delay_cb(uv_timer_t *handle)
+{
+  uv_timer_stop(&main_loop.exit_delay_timer);
+  multiqueue_put(main_loop.fast_events, exit_event, 0);
+}
+
 static void exit_event(void **argv)
 {
+  if (exit_need_delay) {
```

Comment:
There is already `process_is_tearing_down`. But it's not global, and semantics are too different :(

Wonder if we can avoid a new global `exit_need_delay`. I noticed that, although `os_exit` calls `event_teardown`,   `process_teardown` is already waiting for events to drain: https://github.com/neovim/neovim/blob/f50135a32e11c535e1dc3a8e9460c5b4e640ee86/src/nvim/event/process.c#L141-L142

Could we do a similar check here?

Reply:
I tried but it did not work (other stalls occurred). I think that if `exit_event()` is called from a source other than `process_close_handles()` and is delayed, the delayed `exit_event()` will be executed before `main_loop.events` is processed, resulting in an infinite loop.